File: src/api/risk_map.ts
export interface RiskMapNodeDetailsData {
  vulnName?: string | null;
  owasp?: string | null;
  cwe?: string | null;
  cvss?: number | null;
  description?: string | null;
  sastDesc?: string | null;
  sastCode?: string | null;
  dastDesc?: string | null;
  method?: string | null;
  path?: string | null;
  source_file?: string | null;
  line_number?: number | null;
}

export interface RiskMapNodeData {
  id: string;
  label: string;
  type: string;
  status: 'safe' | 'warning' | 'critical' | 'info';
  icon: string;
  x: number;
  y: number;
  score: number;
  severity?: string | null;
  confidence_score?: number | null;
  confidence_level?: string | null;
  verification_status?: string | null;
  source?: string | null;
  project_id: number;
  details?: RiskMapNodeDetailsData | null;
}

export interface RiskMapEdgeData {
  id: string;
  source: string;
  target: string;
  type: string;
  status: 'safe' | 'warning' | 'critical';
  label?: string | null;
}

export interface RiskMapSummaryData {
  total_nodes: number;
  total_edges: number;
  critical_findings: number;
  high_findings: number;
  medium_findings: number;
  low_findings: number;
  info_findings: number;
  overall_risk_score: number;
  overall_risk_grade: string;
}

export interface RiskMapGraphResponseData {
  project_id: number;
  project_name: string;
  nodes: RiskMapNodeData[];
  edges: RiskMapEdgeData[];
  summary: RiskMapSummaryData;
}

const API_BASE_URL = (import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000').replace(/\/$/, '');

export const fetchProjectRiskMap = async (
  projectId: number,
  params?: {
    severity?: string;
    verification_status?: string;
    vulnerability_type?: string;
    min_risk_score?: number;
  }
): Promise<RiskMapGraphResponseData> => {
  const query = new URLSearchParams();
  if (params?.severity) query.append('severity', params.severity);
  if (params?.verification_status) query.append('verification_status', params.verification_status);
  if (params?.vulnerability_type) query.append('vulnerability_type', params.vulnerability_type);
  if (params?.min_risk_score !== undefined) query.append('min_risk_score', String(params.min_risk_score));

  const queryString = query.toString() ? `?${query.toString()}` : '';
  const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}/risk-map${queryString}`);
  if (!response.ok) {
    const errText = await response.text();
    throw new Error(`Failed to load Risk Map (${response.status}): ${errText}`);
  }
  return response.json();
};

File: backend/app/schemas/risk_map.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class RiskMapNodeDetails(BaseModel):
    vulnName: Optional[str] = None
    owasp: Optional[str] = None
    cwe: Optional[str] = None
    cvss: Optional[float] = None
    description: Optional[str] = None
    sastDesc: Optional[str] = None
    sastCode: Optional[str] = None
    dastDesc: Optional[str] = None
    method: Optional[str] = None
    path: Optional[str] = None
    source_file: Optional[str] = None
    line_number: Optional[int] = None


class RiskMapNode(BaseModel):
    id: str
    label: str
    type: str  # 'PROJECT', 'API', 'ENDPOINT', 'FINDING', 'VULNERABILITY', 'FILE'
    status: str  # 'safe', 'warning', 'critical', 'info'
    icon: str
    x: float
    y: float
    score: float
    severity: Optional[str] = None
    confidence_score: Optional[int] = None
    confidence_level: Optional[str] = None
    verification_status: Optional[str] = None
    source: Optional[str] = None
    project_id: int
    details: Optional[RiskMapNodeDetails] = None


class RiskMapEdge(BaseModel):
    id: str
    source: str
    target: str
    type: str  # 'PROJECT_API', 'API_ENDPOINT', 'ENDPOINT_FINDING', 'FINDING_VULN', 'FINDING_FILE'
    status: str  # 'safe', 'warning', 'critical'
    label: Optional[str] = None


class RiskMapSummary(BaseModel):
    total_nodes: int
    total_edges: int
    critical_findings: int
    high_findings: int
    medium_findings: int
    low_findings: int
    info_findings: int
    overall_risk_score: int
    overall_risk_grade: str


class RiskMapGraphResponse(BaseModel):
    project_id: int
    project_name: str
    nodes: List[RiskMapNode]
    edges: List[RiskMapEdge]
    summary: RiskMapSummary
